### PR TITLE
fixes #858 repost

### LINKED ISSUE #858: Document Specific ranges against regex'd TF fields should match against full value as well as tokens
Shard ranges will match against the full value because the FI holds the full value and the tokens, but the TF only holds the tokens. The behavior should be consistent regardless of the range.

### PR DESCRIPTION
**Pull request by FineAndDandy**

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java:35]: It seems to me that we are doing a lot of work to try and keep things sorted when all we are doing is aggregating these into a Document.  Would a better approach be to modify the aggregator to simply do the normalization?

File: warehouse/query-core/src/main/java/datawave/query/iterator/EventFieldIterator.java
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/iterator/EventFieldIterator.java:63]: Should be `compareTo() < 0`

File: warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java
Comment (apmoriarty) [inline warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java:64]: the consistency in these two constructors is killing me.

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java:62]: that's fair, I will create a new document that I pass to apply() then merge it into doc.

File: warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java:38]: This is the same hard coded value as the AttributeFactory cache.

File: warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java:35]: I originally started there, but because it has a 1:many relationship due to normalizers it was going to require interface changes to `IdentityAggregator`. We can still go with that if you prefer.

File: warehouse/query-core/src/main/java/datawave/query/iterator/EventFieldIterator.java
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/iterator/EventFieldIterator.java:63]: yep

File: warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java
Comment (ivakegg) [reply warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java:35]: If it makes the code simpler, then I am all for it.  We have added several new classes here whereas an api change and one simpler extension would be easier to understand and maintain.  I guess the question is how many other places will this affect my changing the interface to return 1:many relationships.

Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/iterator/logic/EventFieldNormalizingIterator.java:35]: It requires changing the parserFieldValues() to return a list rather than a single value, then updating all current methods to return singletons.

File: warehouse/query-core/src/main/java/datawave/query/iterator/EventFieldIterator.java
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/iterator/EventFieldIterator.java:107]: be -> been

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java:60]: The valid bytes in cf.getBackingArray() only go through cf.length().  The backing array could be longer.

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java:83]: Seems to me this whole method could be replaced with simply returning cq.  If you need a copy of the bytes then return new ArrayByteSequence(Arrays.copyOf(cq.getBackingArray(), cq.length()), 0, cq.length()).

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java:130]: This is not actually doing any sorting.

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/iterator/EventFieldIterator.java
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/iterator/EventFieldIterator.java:107]: got it

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java:60]: good catch

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java:83]: you are right, there is no reason to do this. I don't think there is any reason not to just return the cq.

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java
Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/EventFieldAggregator.java:130]: fixed docs

**Review by apmoriarty — DISMISSED** [review]
I think all comments have been addressed

**Review by ivakegg — DISMISSED** [review]

**Comment by FineAndDandy** [comment]
what is this waiting on for merge?

**Comment by drewfarris** [comment]
Let's retarget this to a newer branch, @ivakegg will be creating a release/version3.2 branch

**Comment by FineAndDandy** [comment]
I will retarget this against release/version3.2

**Comment by FineAndDandy** [comment]
I'm working on resolving conflicts. Having some issues with unit tests after the merge

**Comment by FineAndDandy** [comment]
we will need to wait for https://github.com/NationalSecurityAgency/datawave-metadata-utils/pull/7 to be merged and available in release/version3.2 for this ticket to be merged.

**Comment by FineAndDandy** [comment]
this is dependent on #1104

**Review by apmoriarty — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/TLDEventFieldAggregator.java
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/functions/TLDEventFieldAggregator.java:28]: I think this is making an assumption about how we create UIDs.  If we turned on the timestamp portion of the UID then this would break or if we decide to substitute an alternate UID mechanism as it is currently pluggable.  Can we simply look for the last dot and strip that off?

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/functions/TLDEventFieldAggregator.java
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/TLDEventFieldAggregator.java:28]: What if there's more than one dot.

Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/TLDEventFieldAggregator.java]: I updated the code to use UID to get the base UID

Comment (FineAndDandy) [reply warehouse/query-core/src/main/java/datawave/query/jexl/functions/TLDEventFieldAggregator.java]: Generally this query logic is probably not configured right if the UID scheme changes

**Review by ivakegg — APPROVED** [review]

**Review by apmoriarty — APPROVED** [review]

**Pull request merged**
